rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,770 @@
+encapsulation here. If need be we can reference fields that are usually hidden,
+without an error from the compiler.
+
+In addition to lvalues an `MODIFIES_RANGE` can also be terminated with a more
+complex slice expressions as the last projection. This only applies to `*mut`
+pointers to arrays. For instance this is needed for `Vec::truncate` where all of
+the the latter section of the allocation is assigned (dropped).
+
+```rs
+impl<T> Vec<T> {
+  #[modifies(self.buf.ptr.pointer.pointer[len..])]
+  fn truncate(&mut self, len: usize) {
+    ...
+  }
+}
+```
+
+`[..]` denotes the entirety of an allocation, `[i..]`, `[..j]` and `[i..j]` are
+ranges of pointer offsets[^slice-exprs]. The slice indices are offsets with sizing `T`, e.g.
+in Rust `p[i..j]` would be equivalent to
+`std::slice::from_raw_parts(p.offset(i), i - j)`. `i` must be smaller or equal
+than `j`.
+
+A `#[frees(CONDITION, LVALUE)]` clause works similarly to `modifies` but denotes
+memory that is deallocated. It does not admit slice syntax, only lvalues.
+
+[^inferred-footprint]: While inferred memory footprints are sound for both safe
+    and unsafe Rust certain features in unsafe rust (e.g. `RefCell`) get
+    inferred incorrectly and will lead to a failing contract check.
+
+### History Variables
+
+Kani's contract language contains additional support for reasoning how the value
+of mutable memory changes. One case where this is necessary is whenever
+`ensures` needs to reason about state before the function call. By default it
+only has access to state after the call completes, which will be different if
+the call mutates memory.
+
+Returning to our `pop` function from before we may wish to describe in which
+case the result is `Some`. However that depends on whether `self` is empty
+*before* `pop` is called. To do this Kani provides the `old(EXPR)` pseudo
+function (see [this section](#open-questions) about a discussion on naming),
+which evaluates `EXPR` before the call (e.g. to `pop`) and makes the result
+available to `ensures`. It is used like so:
+
+```rs
+impl<T> Vec<T> {
+  #[kani::ensures(old(self.is_empty()) || result.is_some())]
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+`old` allows evaluating any (side-effect free, see [here](#changes-to-other-components)) Rust expression.
+The borrow checker enforces the result of `old` cannot observe the mutations
+from e.g. `pop`, as that would defeat the purpose. If `your` expression in `old`
+returns borrowed content, make a copy instead (using e.g. `clone()`).
+
+Note also that `old` is syntax, not a function and implemented as an extraction
+and lifting during code generation. It can reference e.g. `pop`'s arguments but
+not local variables. Compare the following
+
+**Invalid ❌:** `#[kani::ensures({ let x = self.is_empty(); old(x) } || result.is_some())]`</br>
+**Valid ✅:** `#[kani::ensures(old({ let x = self.is_empty(); x }) || result.is_some())]`
```

Comment:
That is also invalid, but for other reasons.  The `let` itself is the problem. If you call `old` with a locally bound value that'll be an error (`x` is unbound). Actually in the worst case it might even be silently breaking scope, because if the function has an argument `x` then the extraction *could* succeed but `old` would not use the local `let` bound `x` but `x` the function argument. Very confusing. I will note this issue down in open questions.

I can add that as an additional example though.

Reply:
oh... I think I'm confused. So old isn't creating a temporary variable to hold the value instead it is just placing the code before the call?

Reply:
No it is creating a temporary variable which is assigned to the code inside `old`. But if that references locally bound variables they won't be in scope, because only what is inside `old` gets lifted, not its context.